dali/base/dautils.cpp

```diff
@@ -3083,3 +3113,262 @@ bool traceAllTransactions(bool on)
     return ret;
 }
 
+const char* LockDataFieldNames[] = { "@XPath", "@EPCount", "@EP", "@connId", "@sessId", "@mode", "@timeLocked" };
+
+extern da_decl const char* getLockDataFieldName(LockDataField feild)
+{
+    return LockDataFieldNames[feild];
+}
+
+void CLockDataHelper::serializeLockData(CheckedCriticalSection &crit, unsigned critTimeout, const char *xpath, ConnectionInfoMap &connectionInfo, CMessageBuffer &mb)
+{
+    unsigned epCount=0;
+    MemoryBuffer locks;
+    UInt64Array keys;
+    {
+        CHECKEDCRITICALBLOCK(crit, critTimeout);
+        HashIterator iter(connectionInfo);
+        ForEach(iter)
+        {
+            IMapping &imap = iter.query();
+            LockData *lD = connectionInfo.mapToValue(&imap);
+            keys.append(* ((ConnectionId *) imap.getKey()));
+            locks.append(sizeof(LockData), lD);
+            epCount++;
+        }
+    }
+    mb.append(xpath);
+    mb.append(epCount);
+    if (epCount < 1)
+        return;
+
+    ForEachItemIn(k, keys)
+    {
+        LockData lD;
+        memcpy(&lD, ((const byte *)locks.toByteArray())+k*sizeof(LockData), sizeof(LockData));
+
+        StringBuffer sessEpStr;
+        mb.append(querySessionManager().getClientProcessEndpoint(lD.sessId, sessEpStr).str());
+        mb.append(lD.sessId);
+        mb.append(keys.item(k));
+        mb.append(lD.mode);
+        mb.append(lD.timeLockObtained);
+    }
+}
+
+IPropertyTreeIterator *deserializeLockDataIterator(MemoryBuffer &mb)
+{
+    class CLockDataIterator: public CInterface, implements IPropertyTreeIterator
+    {
+        Owned<IPropertyTree> cur;
+
+        IPropertyTree *deserializeLockData(MemoryBuffer &mb)
+        {
+            StringAttr xPath;
+            mb.read(xPath);
+
+            IPropertyTree *lock = createPTree("lock");
+            lock->setProp(getLockDataFieldName(LDFXPath), xPath.get());
+
+            unsigned epCount;
+            mb.read(epCount);
+            lock->setPropInt(getLockDataFieldName(LDFEPCount), epCount);
+            if (epCount < 1)
+                return lock;
+
+            for (unsigned l = 0; l < epCount; l++)
+            {
+                StringAttr ep;
+                __int64 connId, sessId;
+                unsigned mode, timeLockObtained;
+                mb.read(ep);
+                mb.read(sessId);
+                mb.read(connId);
+                mb.read(mode);
+                mb.read(timeLockObtained);
+
+                IPropertyTree* sub = createPTree("EP");
+                sub->setProp(getLockDataFieldName(LDFEP), ep.get());
+                sub->setPropInt64(getLockDataFieldName(LDFconnId), connId);
+                sub->setPropInt64(getLockDataFieldName(LDFsessId), sessId);
+                sub->setPropInt(getLockDataFieldName(LDFmode), mode);
+                sub->setPropInt(getLockDataFieldName(LDFtimeLocked), timeLockObtained);
+                lock->addPropTree("EP", sub);
```

Comment:
Hm, this is another layer of abstraction.. create IPT's to encode LockInfo ..
The serialized info. should be deserialized to the same format as the original.
I see you have 'LockData' in dautils.hpp - so not sure why you are not deserializing to this exposed data type.
It would be more efficient and less obscure.
Would still need to iterate over those though.